elastic/kibana
# [Rule Registry] Initialization errors should surface as a framework execution error.

Resolves https://github.com/elastic/kibana/issues/124159

## Summary

Bubbles up rule registry initialization and write disabled errors to the alerting framework. Previously, when rule registry initialization errors occur, the error is logged and alerts-as-data writing is disabled but no other indication is made, making it hard to debug why alerts as data are not being written.

Once rule registry writing is disabled (either via Kibana config or on initialization error), every rule execution will end in an error.

## To Verify

1. Start up ES and Kibana and create a rule that writes alerts as data and will write with every execution. I used a Metric Threshold rule (and modified the settings to query over the event log index). Verify that the rule runs and alerts are written out.
2. Go to `x-pack/plugins/rule_registry/common/assets/field_maps/technical_rule_field_map.ts` and modify one of the fields so that the type is no longer consistent with the previous type. I changed `Fields.ALERT_SEVERITY` to type `date`. Save and let Kibana restart.
3. The next time the rule runs and tries to write an alert, verify that you see the initialization error logged AND that this rule execution ends in a failure. Here is what my server log looks like:
```
[2022-06-15T08:35:20.783-04:00][INFO ][plugins.ruleRegistry] Installing namespace-level resources and creating concrete index for .alerts-observability.metrics.alerts-default
[2022-06-15T08:35:20.823-04:00][ERROR][plugins.ruleRegistry] Failed to PUT mapping for alias .alerts-observability.metrics.alerts-default: illegal_argument_exception: [illegal_argument_exception] Reason: mapper [kibana.alert.severity] cannot be changed from type [keyword] to [date]
[2022-06-15T08:35:20.824-04:00][ERROR][plugins.ruleRegistry] RuleDataWriterInitializationError: There has been a catastrophic error trying to install namespace level resources for the following registration context: observability.metrics. This may have been due to a non-additive change to the mappings, removal and type changes are not permitted. Full error: ResponseError: illegal_argument_exception: [illegal_argument_exception] Reason: mapper [kibana.alert.severity] cannot be changed from type [keyword] to [date]
    at RuleDataClient.initializeWriter (/Users/ymao/Code/kibana/x-pack/plugins/rule_registry/server/rule_data_client/rule_data_client.ts:178:17)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at RuleDataClient.getWriter (/Users/ymao/Code/kibana/x-pack/plugins/rule_registry/server/rule_data_client/rule_data_client.ts:143:34)
    at Object.executor (/Users/ymao/Code/kibana/x-pack/plugins/rule_registry/server/utils/create_lifecycle_executor.ts:148:34)
    at TaskRunner.executeRule (/Users/ymao/Code/kibana/x-pack/plugins/alerting/server/task_runner/task_runner.ts:378:30)
    at promiseResult (/Users/ymao/Code/kibana/x-pack/plugins/alerting/server/lib/result_type.ts:47:17)
    at TaskRunner.loadRuleAttributesAndRun (/Users/ymao/Code/kibana/x-pack/plugins/alerting/server/task_runner/task_runner.ts:678:25)
    at errorAsRuleTaskRunResult (/Users/ymao/Code/kibana/x-pack/plugins/alerting/server/task_runner/task_runner.ts:1141:12)
    at TaskRunner.run (/Users/ymao/Code/kibana/x-pack/plugins/alerting/server/task_runner/task_runner.ts:734:56)
    at TaskManagerRunner.run (/Users/ymao/Code/kibana/x-pack/plugins/task_manager/server/task_running/task_runner.ts:306:22)
[2022-06-15T08:35:20.859-04:00][ERROR][plugins.ruleRegistry] The writer for the Rule Data Client for the observability.metrics registration context was not initialized properly, bulk() cannot continue, and writing will be disabled.
[2022-06-15T08:35:20.914-04:00][ERROR][plugins.alerting] Executing Rule default:metrics.alert.threshold:516a0940-eca7-11ec-8578-876ba66bdfab has resulted in Error: There has been a catastrophic error trying to install namespace level resources for the following registration context: observability.metrics. This may have been due to a non-additive change to the mappings, removal and type changes are not permitted. Full error: ResponseError: illegal_argument_exception: [illegal_argument_exception] Reason: mapper [kibana.alert.severity] cannot be changed from type [keyword] to [date]
```

And the Rule run history:
<img width="1089" alt="Screen Shot 2022-06-14 at 9 29 29 AM" src="https://user-images.githubusercontent.com/13104637/173589032-8265f381-c57d-476b-a024-509b29b1c459.png">
4. Verify that every execution after this initial error also ends in an error with this message:
```
[2022-06-15T08:36:21.866-04:00][ERROR][plugins.alerting] Executing Rule default:metrics.alert.threshold:516a0940-eca7-11ec-8578-876ba66bdfab has resulted in Error: Rule registry writing is disabled. Make sure that "xpack.ruleRegistry.write.enabled" configuration is not set to false within "kibana.yml" and that Rule Data Client was initialized successfully.
```
5. Revert the field mapping changes from step 2 and verify that rule execution starts to succeed.
6. Set `xpack.ruleRegistry.write.enabled: false` in your Kibana config and verify that rule execution starts failing each time.
7. Reset `xpack.ruleRegistry.write.enabled: true` and verify that rule execution starts to succeed.
